service/spi/src/test/java/org/kaazing/gateway/service/messaging/collections/MemoryTopicTest.java

```diff
@@ -0,0 +1,175 @@
+        ITopic<String> topic = factory.getTopic("topic");
+        CountDownLatch listenersCalled = new CountDownLatch(1);
+        MessageListener m = message -> listenersCalled.countDown();
+        String name = topic.addMessageListener(m);
+        topic.publish("msg1");
+        listenersCalled.await();
+        topic.removeMessageListener(name);
+        topic.publish("msg2");
+        assertEquals(2, topic.getLocalTopicStats().getPublishOperationCount());
+        assertEquals(1, topic.getLocalTopicStats().getReceiveOperationCount());
+    }
+
+
+    @Test
+    public void shouldAddAndRemoveMessageListenerWithoutDeadlock() throws InterruptedException {
+        ITopic<String> topic = factory.getTopic("topic");
+        CountDownLatch listenersCalled = new CountDownLatch(1);
+        MessageListener m1 = message -> {
+            listenersCalled.countDown();
+        };
+        String name = topic.addMessageListener(m1);
+        MessageListener m2 = message -> {
+            topic.removeMessageListener(name);
+        };
+        topic.addMessageListener(m2);
+        topic.publish("msg1");
+        listenersCalled.await();
+    }
+
+
+    @Test
+    public void shouldAddAndRemoveMessageListenerNoSideEffects() throws InterruptedException {
+        ITopic<String> topic = factory.getTopic("topic");
+        CountDownLatch listenersCalled = new CountDownLatch(2);
+        AtomicBoolean calledM3 = new AtomicBoolean(false);
+        MessageListener m1 = message -> {
+            listenersCalled.countDown();
+            topic.addMessageListener(message1 -> calledM3.set(true));
+        };
+        String name = topic.addMessageListener(m1);
+        MessageListener m2 = message -> {
+            topic.removeMessageListener(name);
+            listenersCalled.countDown();
+        };
+        topic.addMessageListener(m2);
+        topic.publish("msg1");
+        listenersCalled.await();
+        assertTrue("MessageListener m3 must not be called", calledM3.get() == false);
+    }
+
+    @Test
+    public void shouldPubSubFromMessageListeners() {
+        CountDownLatch listenersCalled = new CountDownLatch(1);
+        ITopic<String> topic1 = factory.getTopic("topic1");
+        ITopic<String> topic2 = factory.getTopic("topic2");
+        MessageListener m1 = message -> topic2.publish("Message sent to topic2");
+        MessageListener m2 = message -> listenersCalled.countDown();
+        topic1.addMessageListener(m1);
+        topic2.addMessageListener(m2);
+        topic1.publish("trigger m2");
+    }
+
+
+    @Test
+    public void shouldNotOverflow() throws InterruptedException {
+        ITopic<String> topic = factory.getTopic("topic");
+        CountDownLatch latch = new CountDownLatch(10000);
+        topic.addMessageListener(message -> {
+            latch.countDown();
+            if (latch.getCount() > 0) {
+                topic.publish("Resend: " + message.getMessageObject());
```

Comment:
Again I'm not sure we really need to support this. If we can, it does no harm. When I would be equally happy if we just throw an exception from the publish they attempt to this.